flake8_type_checking/checker.py

```diff
@@ -360,28 +361,147 @@ def import_type(self) -> ImportTypeValue:
         return cast('ImportTypeValue', classify_base(self.full_name.partition('.')[0]))
 
 
-class UnwrappedAnnotation(NamedTuple):
-    """Represents a single `ast.Name` in an unwrapped annotation."""
+class WrappedAnnotation(NamedTuple):
+    """Represents a wrapped annotation i.e. a string constant."""
 
     lineno: int
     col_offset: int
     annotation: str
+    names: set[str]
+    scope: Scope
     type: Literal['annotation', 'alias', 'new-alias']
 
 
-class WrappedAnnotation(NamedTuple):
-    """Represents a wrapped annotation i.e. a string constant."""
+class UnwrappedAnnotation(NamedTuple):
+    """Represents a single `ast.Name` in an unwrapped annotation."""
 
     lineno: int
     col_offset: int
     annotation: str
-    names: set[str]
+    scope: Scope
     type: Literal['annotation', 'alias', 'new-alias']
 
 
+class Symbol(NamedTuple):
+    """Represents an import/definition/declaration of a variable."""
+
+    name: str
+    lineno: int
+    col_offset: int
+    type: Literal['import', 'definition', 'declaration', 'argument']
+    in_type_checking_block: bool
+
+    def available_at_runtime(self, use: HasPosition | None = None) -> bool:
+        """Return whether or not this symbol is available at runtime."""
+        if self.in_type_checking_block or self.type == 'declaration':
+            return False
+
+        # we punt in this case, this is to support some use-cases where
+        # the location of use does not matter, such as __all__
+        if use is None:
+            return True
+
+        if use.lineno < self.lineno:
+            return False
```

Comment:
`use` is the node from where you are doing the name lookup for, so it's looking if there's a matching symbol on or above the line of the use. Technically this check is not completely correct, since there's some out-of-order expressions such as `x if (x := foo()) else None` and `(x for x in y)` where the definition can occur lexicographically after the use, but I doubt the number of false positives/negatives this will cause is high enough to justify the extra complication, although maybe there's a simple way of handling this (such as attaching the definition to the start of the expression rather than the actual location of the definition).

That being said I might have forgotten to count walrus assignment as definitions, which is something I definitely should fix, if it is the case. I probably should also deal with generator expressions a bit better, so the loop variables get added to the scope.

Reply:
I did forget walrus assignments, but they're now being handled. I also improved scoping for out of order expressions.

`use` makes more sense in the context of `Scope.lookup` on the `Symbol` itself it is admittedly a little bit more confusing. Maybe I should move this method to `Scope`, since `Scope.lookup` is arguably the only place where it makes sense to call `available_at_runtime`.

Reply:
Yeah, let's just leave it this way for now.